### PR TITLE
Added validation for reducing participants' contest duration below base value

### PR DESCRIPTION
**Pull request by jicata**
Restrict contest duration time reduction to not reduce participants total duration below contest base duration - closes https://github.com/SoftUni-Internal/suls-issues/issues/3761

### CODE REVIEW

File: Open Judge System/Web/OJS.Web/App_GlobalResources/Areas/Administration/Contests/ContestsControllers.bg.resx [hunk 1]
```diff
@@ -150,6 +150,9 @@
   <data name="Duration_invalid_format" xml:space="preserve">
     <value>Продължителността трябва да е във формат "hh:mm"</value>
   </data>
+  <data name="Failed_to_update_participants_duration" xml:space="preserve">
+    <value>&lt;br /&gt;ВНАМНИЕ: Продължителността на състезанието на следните потребители не беше обновена, защото продължителността им би паднала под базовата продължителност на състезанието: {0}</value>
```
Comment (gogo4ds) [inline Open Judge System/Web/OJS.Web/App_GlobalResources/Areas/Administration/Contests/ContestsControllers.bg.resx]: shoul be "ВНИМАНИЕ"

Comment (Innos) [inline Open Judge System/Web/OJS.Web/App_GlobalResources/Areas/Administration/Contests/ContestsControllers.bg.resx]: Typo, should be "ВНИМАНИЕ"

File: Open Judge System/OJS.Services/Data/Participants/ParticipantsDataService.cs [hunk 2]
```diff
@@ -4,7 +4,7 @@
     using System.Data.Entity;
     using System.Linq;
     using System.Linq.Expressions;
-
+    using EntityFramework.Extensions;
```
Comment (Innos) [inline Open Judge System/OJS.Services/Data/Participants/ParticipantsDataService.cs]: Split usings

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs [hunk 3]
```diff
@@ -363,19 +364,34 @@ public ActionResult ChangeActiveParticipantsEndTime(ChangeTimeForParticipantsVie
...
-                model.TimeInMinutes,
-                model.ParticipantsCreatedAfterDateTime.Value,
-                model.ParticipantsCreatedBeforeDateTime.Value);
-
+            var notUpdatedUsers = 
+                this.participantsBusiness.UpdateContestEndTimeForAllParticipantsByContestByParticipantContestStartTimeRangeAndTimeIntervalInMinutes(
+                    model.ContesId, 
+                    model.TimeInMinutes,
+                    model.ParticipantsCreatedAfterDateTime.Value,
+                    model.ParticipantsCreatedBeforeDateTime.Value);
```
Comment (Innos) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs]: Materialize the collection.

File: Open Judge System/OJS.Services/Business/Participants/ParticipantsBusinessService.cs [hunk 4]
```diff
@@ -67,24 +67,38 @@ public bool CanCompeteByContestAndUserId(Contest contest, string userId)
...
             this.participantsData.Update(
-                p =>
-                    p.ContestStartTime >= contestStartTimeRangeStart &&
-                    p.ContestStartTime <= contestStartTimeRangeEnd &&
-                    p.ContestId == contestId &&
-                    p.IsOfficial &&
-                    p.Contest.Type == ContestType.OnlinePracticalExam,
-                p => new Participant() { ContestEndTime = SqlFunctions.DateAdd(
+                participantsInTimeRange
+                .Where(p => SqlFunctions.DateAdd("minute", minutes, p.ContestEndTime) >=
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Services/Business/Participants/ParticipantsBusinessService.cs]: p => on new line

File: Open Judge System/OJS.Services/Business/Participants/ParticipantsBusinessService.cs [hunk 5]
```diff
@@ -67,24 +67,38 @@ public bool CanCompeteByContestAndUserId(Contest contest, string userId)
...
+                    SqlFunctions.DateAdd("minute", contestTotalDurationInMinutes, p.ContestStartTime)),
+                p => new Participant()
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Services/Business/Participants/ParticipantsBusinessService.cs]: tab

**Comment by gogo4ds** [comment]
I added extra check for displaying only the problems of the Participant to ensure that while the contest is open for Compete no regular user can see all the problems https://github.com/SoftUni-Internal/OpenJudgeSystem/pull/188/commits/718412448178c0c8567522d0be7429d31104c0db

**Pull request merged**